lighthouse-viewer/app/src/viewer-ui-features.js

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* global ReportUIFeatures, ReportGenerator */
+
+/**
+ * Extends ReportUIFeatures to add an (optional) ability to save to a gist and
+ * generates the saved report from a browserified ReportGenerator.
+ */
+class ViewerUiFeatures extends ReportUIFeatures {
+  /**
+   * @param {!DOM} dom
+   * @param {?function(!ReportRenderer.ReportJSON)} saveGistCallback
+   */
+  constructor(dom, saveGistCallback) {
+    super(dom);
+
+    /** @private {?function(!ReportRenderer.ReportJSON)} */
+    this._saveGistCallback = saveGistCallback;
+  }
+
+  /**
+   * @param {!ReportRenderer.ReportJSON} report
+   * @override
+   */
+  initFeatures(report) {
+    super.initFeatures(report);
+
+    // Disable option to save as gist if no callback for saving.
+    if (!this._saveGistCallback) {
+      const saveGistItem = this._dom.find('.lh-export--gist', this._document);
+      saveGistItem.setAttribute('disabled', true);
+    }
+  }
+
+  /**
+   * Returns the html that recreates this report. Uses ReportGenerator
+   * @return {string}
+   * @override
+   */
+  getReportHtml() {
+    return new ReportGenerator().generateReportHtml(this.json);
```

Comment:
> would be be able to just use this in main report? Instead of pulling the outerHTML. That would allow us to get rid of `_resetUIState()` too.

We could consider that in a follow up. I avoided it here because it means ReportGenerator has to be browserified before Lighthouse can be run, taking us back to the report build step we just ditched by getting rid of pre-compiled handlebars templates.

I do have a separate idea for getting rid of `_resetUIState()`, though: basically we'd track which templates had been imported in a JS variable instead of an attribute. `outerHTML` wouldn't care anymore, and the equivalent of `dom.resetTemplates()` could be called in `renderReport` at the same time it does `container.textContent = '';` so that they'd be cloned correctly again.

(though I am personally digging the fact that all our uses of the v2 report just builds the report rendering classes new each time. If we only want to go that way, no need for resetting the `Dom` class, it'll just be a new instance every time)